packages/core/core-flows/src/product/workflows/update-products.ts

```diff
@@ -20,7 +37,87 @@ export const updateProductsWorkflow = createWorkflow(
     input: WorkflowData<WorkflowInput>
   ): WorkflowData<ProductTypes.ProductDTO[]> => {
     // TODO: Delete price sets for removed variants
-    // TODO Update sales channel links
-    return updateProductsStep(input)
+
+    const toUpdateInput = transform({ input }, (data) => {
+      if ((data.input as UpdateProductsStepInputProducts).products) {
+        return (data.input as UpdateProductsStepInputProducts).products.map(
+          (p) => ({
+            ...p,
+            sales_channels: undefined,
+          })
+        )
+      } else {
+        return {
+          selector: (data.input as UpdateProductsStepInputSelector).selector,
+          update: {
+            ...(data.input as UpdateProductsStepInputSelector).update,
+            sales_channels: undefined,
+          },
+        }
+      }
+    })
+
+    const updatedProducts = updateProductsStep(toUpdateInput as any) // TODO: type
+
+    const updatedProductIds = transform({ updatedProducts, input }, (data) => {
+      if (
+        !(data.input as UpdateProductsStepInputSelector).update.sales_channels
+      ) {
+        return []
+      }
+
+      let productIds = data.updatedProducts.map((p) => p.id)
+
+      ;(data.input as UpdateProductsStepInputProducts).products?.forEach(
+        (p) => {
+          if (!p.sales_channels) {
+            // these products don't update sales channels so we don't want to their previous channels
+            productIds = productIds.filter((id) => id !== p.id)
+          }
+        }
+      )
+
+      return productIds
+    })
+
+    const currentLinks = getProductSalesChannelLinkStep({
+      ids: updatedProductIds,
+    })
+
+    detachProductsFromSalesChannelsStep({ links: currentLinks })
+
+    const salesChannelLinks = transform({ input, updatedProducts }, (data) => {
+      if ((data.input as UpdateProductsStepInputSelector).selector) {
+        if (
+          !(data.input as UpdateProductsStepInputSelector).update.sales_channels
+        ) {
+          return []
+        }
+        return data.updatedProducts
+          .map((p) => {
+            return (
+              data.input as UpdateProductsStepInputSelector
+            ).update.sales_channels!.map((channel) => ({
+              sales_channel_id: channel.id,
+              product_id: p.id,
+            }))
+          })
+          .flat()
```

Comment:
take a look at the util `deepFlatMap` that we have.
it might help you reduce the complexity of these .map, .flat.

Reply:
Not sure if it's applicable here since we are merging data from 2 objects